File: solutions/clojure/bob/1/src/bob.clj
(ns bob)
  (require '[clojure.string :as st])

(def responses {:silence "Fine. Be that way!", :question "Sure.", :yelled "Whoa, chill out!", :yelledq "Calm down, I know what I'm doing!", :normal "Whatever."})

(defn is_silence? [s]
  (nil? (re-find #"\S" s))
  )

(defn is_question? [s]
  (= (last (st/trim s)) \?)
  )

(defn is_yelling? [s]
   (and (some? (re-find #"\p{Lu}" s)) (nil? (re-find #"\p{Ll}" s)))
  )

  

(defn response-for [s] 
  (cond
    (is_silence? s)  (responses :silence)
    (and (is_question? s) (is_yelling? s))  (responses :yelledq)
    (is_question? s)  (responses :question)    
    (is_yelling? s)  (responses :yelled)
    :default  (responses :normal)
    )
  
)

File: solutions/julia/bob/1/bob.jl
responses = (
    "Sure.",
    "Whoa, chill out!",
    "Calm down, I know what I'm doing!",
    "Fine. Be that way!",
    "Whatever."
)

function is_shouting(stimulus)
    occursin(r"\p{Lu}", stimulus) && !occursin(r"\p{Ll}", stimulus)
end

function is_silence(stimulus)
    !occursin(r"\S", stimulus)
end

function bob(stimulus)
    if is_silence(stimulus) 
        reaction = responses[4]
    elseif last(strip(stimulus)) == '?'
        is_shouting(stimulus) ? reaction = responses[3] : reaction = responses[1]
    elseif is_shouting(stimulus)
        reaction = responses[2]
    else 
        reaction = responses[5]
    end
end
